pkg/controllermanager/controller/seed/seed_backup_bucket_control.go

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package seed
+
+import (
+	"context"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	gardencoreclientset "github.com/gardener/gardener/pkg/client/core/clientset/versioned"
+	gardencorelisters "github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) backupBucketEnqueue(obj client.Object) {
+	key := client.ObjectKeyFromObject(obj)
+	c.seedBackupBucketQueue.Add(key.String())
```

Comment:
Why are you enqueueing the `BackupBucket` instead of the referenced `Seed` here? I guess this would save some code here.

Reply:
I also thought about that but was then just following this example https://github.com/gardener/gardener/blob/277374cfffbd034648d228b1344f22782904cf06/pkg/gardenlet/controller/seed/seed_extension_control.go#L68. Is there an essential difference I don't see?

Reply:
You are not really following this example. This "seed extension" check is exactly implementing what I'm suggesting: It watches `ControllerInstallation`s and enqueues the referenced seed: https://github.com/gardener/gardener/blob/277374cfffbd034648d228b1344f22782904cf06/pkg/gardenlet/controller/seed/seed_extension_control.go#L36-L42
The reconciler then reconciles the `Seed` and does not know about the `ControllerInstallation` anymore.

You are doing it differently: You are watching `BackupBucket`s and then enqueue them here. Your reconciles then "reconciles" the` BackupBucket`, but the only thing it does it to call another reconciler for reconciling the referenced `Seed`.

I am suggesting to remove this intermediate step because it seems to be not necessary. You can directly enqueue the referenced `Seed` and have only one reconciler for the `Seed` directly, like in the example you sent.

Reply:
Okay sure, I'll change it. I still don't understand the difference to gardener/pkg/gardenlet/controller/seed/seed_extension_control.go. For me it's a mapping in both cases, from Resource A -> B. What the "2nd" controller (reconciling B) knows or doesn't know, shouldn't be relevant.

Reply:
Yes, it's a mapping in both cases, but you currently get an event with the object, then enqueue the object's name, then you read the object by name again, and then you call your actual business logic with some information out of the object.

The "seed extension" check gets an event with the object and immediately extracts the relevant information so that the business logic can be called right away.

Obviously, the result/behaviour of the controller is the same. The proposed approach is just simpler and results in less code.